src/common/cuda/rtc.cc

```diff
@@ -0,0 +1,246 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "mxnet/base.h"
+
+#if MXNET_USE_CUDA
+
+#include <nvrtc.h>
+
+#include <mutex>
+#include <string>
+#include <fstream>
+#include <unordered_map>
+#include <vector>
+
+#include "rtc.h"
+#include "rtc/half-inl.h"
+#include "rtc/util-inl.h"
+#include "rtc/forward_functions-inl.h"
+#include "rtc/backward_functions-inl.h"
+#include "rtc/vectorization-inl.h"
+#include "rtc/special_functions-inl.h"
+#include "rtc/reducer-inl.h"
+#include "utils.h"
+
+
+namespace mxnet {
+namespace common {
+namespace cuda {
+namespace rtc {
+
+std::mutex lock;
+
+namespace util {
+
+std::string to_string(OpReqType req) {
+  switch (req) {
+    case kNullOp:
+      return "OpReqType::kNullOp";
+    case kWriteTo:
+    case kWriteInplace:
+      return "OpReqType::kWriteTo";
+    case kAddTo:
+      return "OpReqType::kAddTo";
+  }
+  LOG(FATAL) << "Unrecognized req.";
+  return "";
+}
+
+}  // namespace util
+
+namespace {
+
+// Obtain compilation log from the program.
+std::string GetCompileLog(nvrtcProgram program) {
+  size_t log_size_including_null;
+  NVRTC_CALL(nvrtcGetProgramLogSize(program, &log_size_including_null));
+  // For most std::string implementations, this is probably 1 char bigger than needed.  OK though.
```

Comment:
Why is it needed?

"The elements of a basic_string are stored contiguously, that is, for a basic_string s, &*(s.begin() + n) == &*s.begin() + n for any n in [0, s.size()), or, equivalently, a pointer to s[0] can be passed to functions that expect a pointer to the first element of a null-terminated (since C++11)CharT[] array."

https://en.cppreference.com/w/cpp/string/basic_string

Reply:
I wrote the code awhile back, but I think I see the logic behind it.  Forgive me, but I prefer to think of it by way of a concrete example:

Let's say the compile log had 100 characters. The nvrtcGetProgramLogSize() call would report a size of 101 and the nvrtcGetProgramLog() call would copy 101 characters to the pointer address passed to it.  The 101 characters are the 100 characters of the log plus a null byte.  This approach is targeting a C interface- you're expected to malloc an array of 101 bytes (the length reported to you) and all is well.

Here though, we want the program log ultimately to look like std::string of 100 characters.  However, do we want to allocate the string of size 100, then hope we can copy 101 bytes into it?  I think most implementations of std::string will add 1 to the size you give at string construction time, because they want an efficient implementation of string::c_str().  So it would probably work, but why risk it?  Basically, we allocate the string with length 101, let nvrtcGetProgramLog() copy the 101 bytes into it, then resize the string to length 100 so we don't have the null character messing up any print or other string manipulations we do down the road.

Reply:
Thank you for clarifying @DickJC123.

> I think most implementations of std::string will add 1 to the size you give at string construction time, because they want an efficient implementation of string::c_str(). So it would probably work, but why risk it?

As of C++11, the standard guarantees that "a pointer to s[0] can be passed to functions that expect a pointer to the first element of a null-terminated CharT[] array" according to https://en.cppreference.com/w/cpp/string/basic_string. Ie the pointer returned by `s.c_str()` must be the same as as the pointer to `s[0]`. So there would be no risk in allocating the string of true size and avoid the extra resize.

Reply:
I studied your link more carefully and think my post above missed your point.  Studying it further, I agree now that the following would work under C++11:
```
std::string GetCompileLog(nvrtcProgram program) {
  size_t log_size_including_null;
  NVRTC_CALL(nvrtcGetProgramLogSize(program, &log_size_including_null));
  std::string log(log_size_including_null-1, '\0');
  // Room for terminating null character ensured since C++11
  NVRTC_CALL(nvrtcGetProgramLog(program, &log[0]));
  return log;
}
```
I guess with the version currently in the PR, I was being conservative, not wanting to research the level of C++11 support in all the compilers MXNet was going to see.  Do you see any difference between &log[0] and log.data() in terms of ensuring the compiler adheres to its contract of providing size()+1 bytes of contiguous storage?

Reply:
Actually, I see that the non-const version of data() is only available beginning with C++17.  What's the minimum C++ version that this code needs to work against?  Assuming it's C++11 and we go with &log[0] above, do you think that the array-size guarantee of size()+1 is there if the base data address is obtained via &log[0]?  While the Microsoft online documentation generally mirrors the cppreference link you posted, I couldn't find the text  "pointer to s[0] can be passed to functions that expect a pointer to the first element of a null-terminated (since C++11) CharT[] array" in anything from Microsoft, e.g. https://docs.microsoft.com/en-us/cpp/standard-library/basic-string-class?view=vs-2019

Reply:
@DickJC123 If you read the `data` and `operator[]` paragraphs from the Microsoft documentation that you linked, you will see that they do actually say the same thing (just in a more convoluted way):
 - from `data`: `Prior to C++11, data didn't guarantee the returned string was null-terminated. Since C++11, data and c_str both return a null-terminated string, and are effectively the same.`
 - from `operator[]`: ` An invalid index passed to operator[] results in undefined behavior, but the index equal to the length of the string is a valid index for const strings and the operator returns the null character when passed this index.`

Reply:
I was assuming we wouldn't be using data() or c_str(), since they return a const pointer (data() has a non-const starting with C++17).  I was worried about a clever compiler that ensured the contiguous size()+1 layout and presence of the null byte lazily upon the user calling data() or c_str().  So assuming we access a non-const pointer to the beginning of the string via &log[0], do you feel Microsoft describes that the size()+1 layout exists at that point?

Reply:
We require C++17 for all CPU code (which should include the RTC "managamenet" code):

https://github.com/apache/incubator-mxnet/blob/6bb3d724189e1d5727fc7b2b41cab3863294ff99/CMakeLists.txt#L9-L12

though the CUDA code is compiled with C++14 due to lack of C++17 support in Cuda 10

https://github.com/apache/incubator-mxnet/blob/6bb3d724189e1d5727fc7b2b41cab3863294ff99/CMakeLists.txt#L118-L120

Reply:
You might be surprised to hear that I'm actually ambivalent which way we go here.  The way I see it, the two approaches are:

1. Existing approach- works, more conservative coding style, has unneeded resize() (2 lines)
2. Proposed approach- simpler, small bit faster due to eliminating resize(), closer to 'bleeding edge' of C++11 standard

I propose we go with approach 2, without the resize(), as suggested by @leezu.  However, let's leave in the comment that I included in my code snippet above.  If the RTC ever starts segfaulting, and I'm pulled in to debug it, the comment would be a sufficient reminder to me to revert to the former approach to eliminate compiler standard non-compliance as a possible cause.